File: OrbedenEditor/Src/Editor/Panels/BuildGamePanel.cpp
#include "Editor/Panels/BuildGamePanel.h"

#include "Editor/EditorSystem.h"
#include "Editor/Panels/EditorPanelRegistry.h"

#include <imgui.h>
#include <string>

BuildGamePanel::BuildGamePanel(EditorSystem& owner)
    : editor(owner)
{
    info.id = "build_game";
    info.title = "Build Game";
    info.defaultVisible = true;
    info.defaultSize = { 360.0f, 180.0f };
    info.defaultDock = PanelDockPlacement::Floating;
    info.order = 110;
}

//获取面板信息
const EditorPanelInfo& BuildGamePanel::GetPanelInfo() const
{
    return info;
}

//绘制面板内容
void BuildGamePanel::DrawPanel()
{
    if (editor.HasProject())
    {
        ImGui::Text("Name: %s", editor.GetProjectName().c_str());
        ImGui::TextWrapped("Root: %s", editor.GetProjectRoot().c_str());
        ImGui::TextWrapped("World: %s", editor.GetStartupWorldPath().c_str());
        ImGui::TextWrapped("Script: %s", editor.GetProjectScriptRootPath().c_str());
        ImGui::TextWrapped("Managed: %s", editor.GetProjectManagedRootPath().c_str());
        std::string nativeRoot = editor.GetProjectNativeRootPath();
        ImGui::TextWrapped("Native: %s", nativeRoot.empty() ? "<disabled>" : nativeRoot.c_str());
    }
    else
    {
        ImGui::TextUnformatted("No project loaded.");
    }

    ImGui::Separator();

    if (!editor.HasProject())
    {
        ImGui::BeginDisabled();
    }

    if (ImGui::Button("Build Game C#"))
    {
        editor.RequestBuildScripts();
    }

    ImGui::SameLine();
    if (ImGui::Button("Build Game C++"))
    {
        editor.RequestBuildNative();
    }

    ImGui::SetNextItemWidth(180.0f);
    if (ImGui::BeginCombo("Target Platform", editor.GetSelectedPlayerTargetPlatformName()))
    {
        int32 selectedIndex = editor.GetSelectedPlayerTargetPlatformIndex();
        for (int32 index = 0; index < editor.GetPlayerTargetPlatformCount(); ++index)
        {
            bool selected = index == selectedIndex;
            if (ImGui::Selectable(editor.GetPlayerTargetPlatformName(index), selected))
            {
                editor.SetSelectedPlayerTargetPlatformIndex(index);
            }

            if (selected)
            {
                ImGui::SetItemDefaultFocus();
            }
        }

        ImGui::EndCombo();
    }

    ImGui::SameLine();
    if (ImGui::Button("Build Player"))
    {
        editor.RequestBuildPlayer();
    }

    if (!editor.HasProject())
    {
        ImGui::EndDisabled();
    }

    const std::string& status = editor.GetProjectStatusText();
    if (!status.empty())
    {
        ImGui::Separator();
        ImGui::TextWrapped("%s", status.c_str());
    }
}

ORBEDEN_REGISTER_EDITOR_PANEL(BuildGamePanel)

File: OrbedenEditor/Managed/Orbeden.Editor/Panels/BuildGamePanel.cs
using Orbeden;

namespace OrbedenEditor;

/// <summary>显示项目构建信息并调用原生构建服务。</summary>
internal sealed class BuildGamePanel : EditorPanel
{
    public override EditorPanelInfo Info => new("build_game", "Build Game", true,
        new vector2(360, 180), PanelDockPlacement.Floating, 0.25f, 110);

    /// <summary>绘制项目路径、构建目标和构建状态。</summary>
    protected override void DrawContent(EditorPanelContext context)
    {
        string root = EditorApplication.GetProjectText(EditorProjectField.Root);
        if (root.Length == 0) EditorGUI.Label("No project loaded.");
        else
        {
            foreach (EditorProjectField field in new[] { EditorProjectField.Name, EditorProjectField.Root,
                EditorProjectField.Content, EditorProjectField.World, EditorProjectField.Managed, EditorProjectField.Native })
                EditorGUI.Label(field + ": " + EditorApplication.GetProjectText(field));
        }
        EditorGUI.Separator();
        EditorGUI.BeginDisabled(root.Length == 0);
        try
        {
            if (EditorGUI.Button("Build Game C#")) EditorApplication.RequestBuild(EditorBuildKind.Scripts);
            EditorGUI.SameLine();
            if (EditorGUI.Button("Build Game C++")) EditorApplication.RequestBuild(EditorBuildKind.Native);

            //显示平台可用性并保存选择
            string[] targets = EditorApplication.GetProjectText(EditorProjectField.PlayerTargets).Split('\0');
            int selected = EditorApplication.GetSelectedPlayerTarget();
            string preview = selected >= 0 && selected * 2 < targets.Length ? targets[selected * 2] : string.Empty;
            if (EditorGUI.BeginCombo("Target Platform", preview))
            {
                try
                {
                    for (int index = 0; index * 2 + 1 < targets.Length; ++index)
                    {
                        EditorGUI.BeginDisabled(targets[index * 2 + 1] != "1");
                        try
                        {
                            if (EditorGUI.Selectable(targets[index * 2], index == selected))
                                EditorApplication.SetSelectedPlayerTarget(index);
                        }
                        finally { EditorGUI.EndDisabled(); }
                    }
                }
                finally { EditorGUI.EndCombo(); }
            }
            EditorGUI.SameLine();
            if (EditorGUI.Button("Build Player")) EditorApplication.RequestBuild(EditorBuildKind.Player);
        }
        finally { EditorGUI.EndDisabled(); }
        string status = EditorApplication.GetProjectText(EditorProjectField.Status);
        if (status.Length == 0) return;
        EditorGUI.Separator();
        EditorGUI.Label(status);
    }
}
